File: Chapter02/scala/packt2/spark/WetSchemaParser.scala
package packt2.spark

import Utilities01.HelperScala
import Utilities02.WetRecord
import org.apache.hadoop.io.Text
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.SparkSession
import Utilities01.HelperScala.createSession
import Utilities02.HelperScala.{sampleWetLoc, extractRawRecords, parseRawWet}

/**
 * Code for parsing .wet files of the WARC corpus
 *
 * @author Phil, https://github.com/g1thubhub
 */
object WetSchemaParser {

  def main(args: Array[String]) = {

    val inputLocWet = sampleWetLoc
    implicit val session: SparkSession = createSession(3, "WET Parser")
    session.sparkContext.setLogLevel("ERROR") // avoids printing of info messages

    val rawRecords: RDD[Text] = extractRawRecords(inputLocWet)
    val wetRecords: RDD[WetRecord] = rawRecords
      .flatMap(parseRawWet(_))

    import session.implicits._
    wetRecords.toDF().printSchema()
    println(s"Total records: ${wetRecords.count()}")

    /*
    val textRecords = wetRecords
      .filter(_.warcType != "warcinfo") // skip meta header info for file
      .toDF()

    textRecords.show(3)
    val wikipediaRecords = textRecords.as[WetRecord].filter(e => e.targetURI.contains("wikipedia"))
    print(wikipediaRecords.count())
    // 1
    val wikipediaTexts = wikipediaRecords.map(_.plainText)
    println(wikipediaTexts.first())
    */

  }
}

File: Chapter02/python/packt2/spark/wet_schema_parser.py
from utilities01_py.helper_python import create_session
from utilities02_py.helper_python import extract_raw_records, parse_raw_wet
from globalp.python.packtg.helper_python_global import sample_wet_loc

if __name__ == "__main__":
    wet_loc = sample_wet_loc
    session = create_session(3, 'Schema Parsing WET')
    session.sparkContext.setLogLevel('ERROR')  # avoids printing of info messages

    raw_records = extract_raw_records(wet_loc, session)
    wet_records = raw_records \
        .flatMap(lambda record: parse_raw_wet(record))

    wet_records.toDF().printSchema()
    print('Total records: ' + str(wet_records.count()))

# text_records = wet_records \
#     .filter(lambda record: record.warc_type != "warcinfo") \
#     .toDF()
#
# text_records.printSchema()
# text_records.show(3)
#
# wiki_records = text_records.rdd.filter(lambda record: 'wikipedia' in record.target_uri)
# print(wiki_records.count())
# # 1
# wikipediaTexts = wiki_records.map(lambda record: record.plain_text)
# print(wikipediaTexts.first())
